src/time_integrator.f90

```diff
@@ -55,23 +63,175 @@ end function constructor
   subroutine step(self, u, v, w, du, dv, dw, dt)
     implicit none
 
-    class(time_intg_t), intent(in) :: self
-    class(field_t), intent(inout) :: u, v, w
-    class(field_t), intent(in) :: du, dv, dw
+    class(time_intg_t), intent(inout) :: self
+    class(field_t), target, intent(inout) :: u, v, w
+    class(field_t), target, intent(in) :: du, dv, dw
 
     real(dp), intent(in) :: dt
 
-    call self%backend%vecadd(dt, du, 1._dp, u)
-    call self%backend%vecadd(dt, dv, 1._dp, v)
-    call self%backend%vecadd(dt, dw, 1._dp, w)
+    ! assign pointer to variables
+    self%curr(1)%ptr => u
+    self%curr(2)%ptr => v
+    self%curr(3)%ptr => w
+
+    ! assign pointer to variables
+    self%deriv(1)%ptr => du
+    self%deriv(2)%ptr => dv
+    self%deriv(3)%ptr => dw
+
+    ! First-order
+    if (self%order .eq. 1) then
+      call self%adams_bashforth_1st(dt)
+
+    ! Second-order
+    else if (self%order .eq. 2) then
+      if (self%istep .eq. 0) then
+        call self%adams_bashforth_1st(dt)
+      else
+        call self%adams_bashforth_2nd(dt)
+      end if
+
+    ! Third-order
+    else if (self%order .eq. 3) then
+      if (self%istep .eq. 0) then
+        call self%adams_bashforth_1st(dt)
+      else if (self%istep .eq. 1) then
+        call self%adams_bashforth_2nd(dt)
+      else
+         call self%adams_bashforth_3rd(dt)
+      end if
+
+    ! Fourth-order
+    else if (self%order .eq. 4) then
+      if (self%istep .eq. 0) then
+        call self%adams_bashforth_1st(dt)
+      else if (self%istep .eq. 1) then
+        call self%adams_bashforth_2nd(dt)
+      else if (self%istep .eq. 2) then
+        call self%adams_bashforth_3rd(dt)
+      else
+        call self%adams_bashforth_4th(dt)
+      end if
+    end if
+
+
+    ! increment step counter
+    self%istep = self%istep + 1;
 
   end subroutine step
 
-  subroutine adams_bashford_1st(vels, olds, coeffs)
-    type(flist_t) :: vels(:), olds(:)
-    real :: coeffs(:)
+  subroutine adams_bashforth_1st(self, dt)
+    class(time_intg_t), intent(inout) :: self
+    real(dp), intent(in) :: dt
+
+    integer :: i
+
+    do i = 1, self%nvars
+      call self%backend%vecadd(dt, self%deriv(1)%ptr, 1._dp, self%curr(1)%ptr)
```

Comment:
I think we should use `i` here instead of `1` when accessing entries of `deriv` and `curr`.